vm/src/vm.rs

```diff
@@ -56,23 +56,26 @@ use crate::sysmodule;
 pub struct VirtualMachine {
     pub builtins: PyObjectRef,
     pub sys_module: PyObjectRef,
-    pub stdlib_inits: RefCell<HashMap<String, stdlib::StdlibInitFunc>>,
     pub ctx: PyContext,
     pub frames: RefCell<Vec<FrameRef>>,
     pub wasm_id: Option<String>,
     pub exceptions: RefCell<Vec<PyBaseExceptionRef>>,
-    pub frozen: RefCell<HashMap<String, bytecode::FrozenModule>>,
-    pub import_func: RefCell<PyObjectRef>,
+    pub import_func: PyObjectRef,
     pub profile_func: RefCell<PyObjectRef>,
     pub trace_func: RefCell<PyObjectRef>,
-    pub use_tracing: RefCell<bool>,
-    pub signal_handlers: RefCell<[PyObjectRef; NSIG]>,
-    pub settings: PySettings,
+    pub use_tracing: Cell<bool>,
     pub recursion_limit: Cell<usize>,
-    pub codec_registry: RefCell<Vec<PyObjectRef>>,
+    pub signal_handlers: Option<RefCell<[PyObjectRef; NSIG]>>,
+    pub state: Arc<PyGlobalState>,
     pub initialized: bool,
 }
 
+pub struct PyGlobalState {
+    pub settings: PySettings,
+    pub stdlib_inits: HashMap<String, stdlib::StdlibInitFunc>,
+    pub frozen: HashMap<String, bytecode::FrozenModule>,
```

Comment:
Maybe the frozen modules should be a global static? This should not change between VMs.

Reply:
Maybe, but it's possible that someone would want to inject a frozen module into one VM to wrap their API or something, but want to use another vm for something different. I've done that myself in another project that uses RustPython.